homeassistant/components/apcupsd/binary_sensor.py

```diff
@@ -1,43 +1,81 @@
 """Support for tracking the online status of a UPS."""
 from __future__ import annotations
 
-import voluptuous as vol
+import logging
 
-from homeassistant.components.binary_sensor import PLATFORM_SCHEMA, BinarySensorEntity
-from homeassistant.const import CONF_NAME
+from homeassistant.components.binary_sensor import (
+    BinarySensorEntity,
+    BinarySensorEntityDescription,
+)
+from homeassistant.config_entries import ConfigEntry
 from homeassistant.core import HomeAssistant
-import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.entity import DeviceInfo
 from homeassistant.helpers.entity_platform import AddEntitiesCallback
-from homeassistant.helpers.typing import ConfigType, DiscoveryInfoType
 
-from . import DOMAIN, KEY_STATUS, VALUE_ONLINE
+from . import DOMAIN, VALUE_ONLINE, APCUPSdData
 
-DEFAULT_NAME = "UPS Online Status"
-PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
-    {vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string}
+_LOGGER = logging.getLogger(__name__)
+_DESCRIPTION = BinarySensorEntityDescription(
+    key="statflag",
+    name="UPS Online Status",
+    icon="mdi:heart",
 )
 
 
-def setup_platform(
+async def async_setup_entry(
     hass: HomeAssistant,
-    config: ConfigType,
-    add_entities: AddEntitiesCallback,
-    discovery_info: DiscoveryInfoType | None = None,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
 ) -> None:
     """Set up an APCUPSd Online Status binary sensor."""
-    apcups_data = hass.data[DOMAIN]
+    data_service: APCUPSdData = hass.data[DOMAIN][config_entry.entry_id]
+
+    # Do not create the binary sensor if APCUPSd does not provide STATFLAG field for us
+    # to determine the online status.
+    if data_service.statflag is None:
+        return
 
-    add_entities([OnlineStatus(config, apcups_data)], True)
+    async_add_entities(
+        [OnlineStatus(data_service, _DESCRIPTION)],
+        update_before_add=True,
+    )
 
 
 class OnlineStatus(BinarySensorEntity):
-    """Representation of an UPS online status."""
+    """Representation of a UPS online status."""
 
-    def __init__(self, config, data):
+    def __init__(
+        self,
+        data_service: APCUPSdData,
+        description: BinarySensorEntityDescription,
+    ) -> None:
         """Initialize the APCUPSd binary device."""
-        self._data = data
-        self._attr_name = config[CONF_NAME]
+        if (serial_no := data_service.serial_no) is not None:
+            self._attr_unique_id = f"{serial_no}_{description.key}"
+        self.entity_description = description
+        self._data_service = data_service
+
+    @property
+    def device_info(self) -> DeviceInfo | None:
```

Comment:
We can set `self._attr_device_info` in the init method instead.

Reply:
Updated!